File: src/main/java/com/github/sorusclient/client/event/EventManager.java
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.event;

import lombok.val;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class EventManager {

    private static final Map<Class<?>, List<Consumer<Object>>> consumers = new HashMap<>();

    public static <T> void register(Class<T> clazz, Consumer<T> consumer) {
        consumers.computeIfAbsent(clazz, k -> new ArrayList<>()).add((Consumer<Object>) consumer);
    }

    public static void call(Object event) {
        val consumers1 = consumers.get(event.getClass());
        if (consumers1 == null) return;

        for (val consumer : consumers1) {
            consumer.accept(event);
        }
    }

}

File: src/main/kotlin/com/github/sorusclient/client/event/EventManager.kt
package com.github.sorusclient.client.event

object EventManager {

    val consumers: MutableMap<Class<*>, MutableList<(Any) -> Unit>> = HashMap()

    @Suppress("UNCHECKED_CAST")
    inline fun <reified T> register(noinline consumer: (T) -> Unit) {
        consumers.computeIfAbsent(T::class.java as Class<*>) { ArrayList() }
            .add(consumer as (Any) -> Unit)
    }

    fun call(event: Any) {
        val consumers = consumers[event.javaClass] ?: return
        for (consumer in consumers) {
            consumer(event)
        }
    }

}